### PR TITLE
GCF: add Java Pub/Sub samples

### PR DESCRIPTION
**Pull request by ace-n**

### CODE REVIEW

File: functions/snippets/pubsub/pom.xml [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>com.example.cloud.functions</groupId>
+  <artifactId>functions-snippets-pubsub</artifactId>
+
+  <parent>
+    <groupId>com.google.cloud.samples</groupId>
+    <artifactId>shared-configuration</artifactId>
+    <version>1.0.15</version>
+  </parent>
```
Comment (kurtisvg) [inline functions/snippets/pubsub/pom.xml:31]:
```suggestion
   <!--
      The parent pom defines common style checks and testing strategies for our samples.
      Removing or replacing it should not affect the execution of the samples in anyway.
    -->
   <parent>
    <groupId>com.google.cloud.samples</groupId>
    <artifactId>shared-configuration</artifactId>
    <version>1.0.15</version>
  </parent>
```

File: functions/snippets/pubsub/src/test/java/com/example/functions/pubsub/PubsubTest.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    responseOut = new StringWriter();
+    writerOut = new BufferedWriter(responseOut);
+    PowerMockito.when(response.getWriter()).thenReturn(writerOut);
+
+    logHandler.clear();
+  }
+
+  @After
+  public void afterTest() {
+    //logHandler.flush();
```
Comment (kurtisvg) [inline functions/snippets/pubsub/src/test/java/com/example/functions/pubsub/PubsubTest.java:85]: Is this not needed? If not can we remove?

File: functions/snippets/pubsub/src/main/java/com/example/functions/pubsub/PublishMessage.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import com.google.pubsub.v1.ProjectTopicName;
+import com.google.pubsub.v1.PubsubMessage;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
```
Comment (kurtisvg) [inline functions/snippets/pubsub/src/main/java/com/example/functions/pubsub/PublishMessage.java:34]: We've decided to recommend [sl4fj for logging](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/master/SAMPLE_FORMAT.md#logging) so we can be consistent everywhere. Would you mind updating to use that instead?

File: functions/snippets/pubsub/pom.xml (hunk 1 above)
Comment (ace-n) [reply functions/snippets/pubsub/pom.xml:31]: Done. (I'll do this for the other samples in a separate PR.)

File: functions/snippets/pubsub/src/test/java/com/example/functions/pubsub/PubsubTest.java (hunk 2 above)
Comment (ace-n) [reply functions/snippets/pubsub/src/test/java/com/example/functions/pubsub/PubsubTest.java:85]: Good catch - removed.

File: functions/snippets/pubsub/src/main/java/com/example/functions/pubsub/PublishMessage.java (hunk 3 above)
Comment (ace-n) [reply functions/snippets/pubsub/src/main/java/com/example/functions/pubsub/PublishMessage.java:34]: Unfortunately `slf4j`'s API itself does **not** work with `TestLogHandler` from `com.google.guava:guava-testlib:28.2-jre`, as it doesn't have `java.util.logging`'s `addHandler` function.  (`TestLogHandler` is used to verify that certain statements were logged.)

Happy to make this change (in another PR if need be) if we can find a workaround. The only one I know of right now is to mock log calls that use `TestLogHandler`. (@eamonnmcmanus had previously voted against this approach though)

Comment (kurtisvg) [reply functions/snippets/pubsub/src/main/java/com/example/functions/pubsub/PublishMessage.java:34]: Sounds good for now, but we should figure out how to handle these in tests or flip back our decision to require it.

@lesv FYI

**Comment by lesv** [comment]
# Java 11
```
------------------------------------------------------------
- testing functions/snippets/pubsub
------------------------------------------------------------
     [java] JVM args ignored when same JVM is used.
     [java] The following classes needed for analysis were missing:
     [java]   makeConcatWithConstants
[ERROR] This use of java/util/logging/Logger.info(Ljava/lang/String;)V might be used to include CRLF characters into log messages [com.example.functions.pubsub.PublishMessage] At PublishMessage.java:[line 56] CRLF_INJECTION_LOGS
[ERROR] This use of java/util/logging/Logger.log(Ljava/util/logging/Level;Ljava/lang/String;Ljava/lang/Throwable;)V might be used to include CRLF characters into log messages [com.example.functions.pubsub.PublishMessage] At PublishMessage.java:[line 72] CRLF_INJECTION_LOGS
[ERROR] Method com.example.functions.pubsub.PublishMessage.service(HttpRequest, HttpResponse) appears to call the same method on the same object redundantly [com.example.functions.pubsub.PublishMessage] At PublishMessage.java:[line 64] PRMC_POSSIBLY_REDUNDANT_METHOD_CALLS
[ERROR] This use of java/util/logging/Logger.info(Ljava/lang/String;)V might be used to include CRLF characters into log messages [com.example.functions.pubsub.SubscribeToTopic, com.example.functions.pubsub.SubscribeToTopic] At SubscribeToTopic.java:[line 40]At SubscribeToTopic.java:[line 38] CRLF_INJECTION_LOGS
[ERROR] Class com.example.functions.pubsub.eventpojos.PubSubMessage does not implement a toString method [com.example.functions.pubsub.eventpojos.PubSubMessage] At PubSubMessage.java:[lines 22-61] IMC_IMMATURE_CLASS_NO_TOSTRING
[ERROR] Tests run: 3, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 1.42 s <<< FAILURE! - in com.example.functions.pubsub.PubsubTest
[ERROR] com.example.functions.pubsub.PubsubTest.functionsPubsubPublish_shouldPublishMessage  Time elapsed: 0.027 s  <<< ERROR!
java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:221)
	at java.base/java.util.Optional.<init>(Optional.java:107)
	at java.base/java.util.Optional.of(Optional.java:120)
	at com.example.functions.pubsub.PubsubTest.functionsPubsubPublish_shouldPublishMessage(PubsubTest.java:98)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:377)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:138)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:465)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:451)

[ERROR] com.example.functions.pubsub.PubsubTest.functionsPubsubSubscribe_shouldPrintPubsubMessage  Time elapsed: 0.02 s  <<< ERROR!
org.mockito.exceptions.misusing.UnfinishedStubbingException:

Unfinished stubbing detected here:
-> at com.example.functions.pubsub.PubsubTest.functionsPubsubPublish_shouldPublishMessage(PubsubTest.java:98)

E.g. thenReturn() may be missing.
Examples of correct stubbing:
    when(mock.isOk()).thenReturn(true);
    when(mock.isOk()).thenThrow(exception);
    doThrow(exception).when(mock).someVoidMethod();
Hints:
 1. missing thenReturn()
 2. you are trying to stub a final method, which is not supported
 3: you are stubbing the behaviour of another mock inside before 'thenReturn' instruction is completed

	at com.example.functions.pubsub.PubsubTest.beforeTest(PubsubTest.java:73)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:377)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:138)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:465)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:451)

[ERROR] Errors:
[ERROR]   PubsubTest.functionsPubsubPublish_shouldPublishMessage:98 » NullPointer
[ERROR]   PubsubTest.beforeTest:73 UnfinishedStubbing
Unfinished stubbing detected here...
[ERROR] Tests run: 3, Failures: 0, Errors: 2, Skipped: 0
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M4:test (default-test) on project functions-snippets-pubsub: There are test failures.
```

**Comment by lesv** [comment]
That looks like 2 very clear errors.  What do you want changed?  All the "Errors" don't stop the build. The NPE however is real and detected at RUNTIME and that is just the kind of error we do want to stop going into our codebase at this time.

**Comment by ace-n** [comment]
@lesv correct - those are **JUnit** errors. (I assume you're referring to `spotbugs` - I've been holding off on those until we iron out the whitelist.)

The `DEBUG` commit is intended to debug them. (I suspect an environment variable is not getting picked up correctly, but the set-up looks correct from what I can see.)

**Comment by ace-n** [comment]
Failure should be resolved now. (I was updating the wrong config file for environment variables.)

**Comment by ace-n** [comment]
Merging this for now - will add any (further) changes recommended in #2630 into a separate PR. 🙂

**Pull request merged**